src/hotspot/share/classfile/classPrinter.hpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_CLASSFILE_CLASSPRINTER_HPP
+#define SHARE_CLASSFILE_CLASSPRINTER_HPP
+
+#include "memory/allStatic.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+class InstanceKlass;
+class Symbol;
+
+// ClassPrinter is intended to be called from findclass/findmethod/findmethod2
+// in debug.cpp (inside a debugger, such as gdb). To avoid deadlocks (as the JVM
+// may be at an arbitrary native breakpoint), ClassPrinter calls
+// ClassLoaderDataGraph::classes_do without holding any locks. The down side is
+// that the printing may proceed while other threads are running, so race conditions
+// are possible. Use with care.
+//
+// If you want to call these functions programmatically, make sure the caller
+// holds the appropriate locks.
+class ClassPrinter : public AllStatic {
+  class KlassPrintClosure;
+
+  static bool matches(const char *pattern, const char *candidate, int p, int c);
+  static bool matches(const char* pattern, Symbol* symbol);
+  static bool matches_klass(const char* pattern, InstanceKlass* ik);
+
+public:
+
+  enum Mode : int {
+    PRINT_METHOD_NAME       = 1 << 0,
+    PRINT_BYTECODE          = 1 << 1,
+    PRINT_BYTECODE_ADDR     = 1 << 2,
+    PRINT_DYNAMIC           = 1 << 3, // extra information for invokedynamic (and dynamic constant ...)
+  };
+
+  // flags must be OR'ed from ClassPrinter::Mode for the next 3 functions
```

Comment:
Can you give us a mode PRINT_ALL that or'es all of them modes together, and make that the default for flags?

Reply:
gdb doesn't support default parameters (I think the info for default params is not available in the debug info), so you need to pass in `flags` explicitly. If you pass 0 for `findclass`, it prints only the class names. If you pass 0 for `findmethod` and `findmethods`, they print only the class names and method names.

Printing everything by default may be too overwhelming.

BTW, I think  `findmethod` vs `findmethods` may not be the right names. The former takes only method name, while the latter also takes a signature. Maybe I should rename the latter to `findmethodex`?